### ISSUE TITLE
[Question] FlexVolume

### ISSUE THREAD

**Issue by batiati** [democratic-csi/democratic-csi#11]
Hi,
First of all, thank you for this great project, it's very handy! 

In order to support some legacy applications, I have to deal with Microsoft Windows nodes on k8s. As you must know, Windows supports few storages options, SMB and iSCSI through [FlexVolume](https://kubernetes.io/docs/concepts/storage/volumes/#flexVolume) drivers, and no NFS support.

I'm considering to implement a CSI daemon-set running on Windows nodes to translate the democratic-csi iSCSI claims to the equivalent of FlexVolume scripts call.

For example, that is the way I have to configure my Windows pod to connect to a iSCSI volume:

```
  volumes:
  - flexVolume:
      driver: microsoft.com/iscsi.cmd
      fsType: NTFS
      options:
        authType: NONE
        chapAuthDiscovery: "false"
        chapAuthSession: "false"
        iqn: iqn.2005-10.org.freenas.ctl:csi-pvc-6933c3a8-e7fd-4e30-80cf-f5e7c6e4fae0-labs
        lun: "0"
        portals: someserver
        targetPortal: somePortal
      secretRef: {}
    name: vol1
```

And that is how I would like to configure it (as any other pod using democratic-csi)

```
  volumes:
      - name: vol1
        persistentVolumeClaim:
          claimName: test-volume
```

**Questions**

Do you see a better option or some limitation on this approach?

Do you have interest to include this kind of feature on democratic-csi? (So I have to choose either to implement it on JS following the project's guidlines or implement using something else ...) 

Regards!

**Comment by travisghansen** [comment]
Thanks for the interest! I’m admittedly ignorant when it comes to how volumes are handled with windows and what the tooling looks like to integrate with k8s/csi. I’d need to do some research to better understand the possible approaches.

Having said that, I’m more than happy to have windows support added/built. There are some good news here:

- implementing a share over smb would be really easy (server side anyway, not sure about nodes)
- it would be so easy I nearly ended up doing it just for fun when I got started with the project (the code was implemented with this in mind)
- even though it’s a unified code/image scenario, there is complete separation (by design) of the controller and node aspects...meaning you could deploy/create an entirely new image/project to handle the node attachment if necessary
- the ‘contract’ for the node operations is very simple and straightforward
- if certain features  aren’t supported by windows (like an online resize or whatever) the configuration files technically expose all the internals of what the driver broadcasts etc so it’s easy to do non standard deployment scenarios 

In short, let’s explore the options jointly and get a working solution one way or another.

**Comment by batiati** [comment]
Great!
I'm going to build a quick proof of concept, and then we decide what to do next.

- Indeed, SMB is very easy to implement on Windows nodes too, but it's not suited to some applications that expect a block device, I had some trouble with it.

- The microsoft/iscsi.cmd FlexVolume implementation is just a bunch of poweshell scripts that follows the FlexVolume call interface, I think that we can modify it to support any feature (snapshots, resize, etc)

Thanks, I'll get in touch soon

**Comment by travisghansen** [comment]
I'm not sure how bridging to `FlexVolume` would work as it relates to `csi`. Meaning, I'm not sure that would even work from a theoretical standpoint.

My guess is we're going to need to go this route: https://github.com/kubernetes-csi/csi-proxy

In short, you install the proxy on your windows nodes, and then the csi node implementation (this project, or a new sister project to this) can then invoke calls to the proxy to mount/format/unmount/etc/etc over grpc.

I'm currently in the `csi-windows` slack channel trying to get some info.

**Comment by batiati** [comment]
> I'm not sure how bridging to FlexVolume would work as it relates to csi. Meaning, I'm not sure that would even work from a theoretical standpoint.

I've read a little about CSI on Windows, we can mount a `named pipe` in the same way we used to do with `unix socks`, replacing `- --server-socket=/csi-data/csi.sock` to pipe's mount path.

```
      containers:
      - args:
        - --csi-version=1.1.0
        - --csi-name=org.democratic-csi.iscsi
        - --driver-config-file=/config/driver-config-file.yaml
        - --log-level=info
        - --csi-mode=node
        - --server-socket=/csi-data/csi.sock
```

You already did most of the hard work with all gRPC stuff, I think that we can just implement a windows-specifc `iscsi.js` backed on the `FlexVolume` powershell scripts; I have tested those from inside the container, it seems to work in the same way `iscsiadm` do.

![image](https://user-images.githubusercontent.com/5728002/92159821-65989600-ee04-11ea-9897-a1c904722327.png)

I know, It may be tricky to put it all working together, but that's it!

Regards

**Comment by travisghansen** [comment]
Well, if I'm understanding things correctly (and your proposed solution) the named pipe won't help because the node container (on windows) cannot run in privileged mode. That's why the csi-proxy project exists so that a un-privileged container can communicate with the host using an already-running process (csi-proxy) over grpc.

I suppose you could *not* run the node process as a container on your nodes and rather run it directly outside the container as a service. At that point you would bind-mount the existing socket from the host-level app into the 'driver' deployment running democratic-csi outside of k8s but still let k8s invoke the grpc calls directly (with system-level access).

So options are like this:

- k8s -> csi grpc call -> un-privileged democratic-csi container -> csi-proxy grpc call -> csi-proxy running privileged directly on the node

OR

- k8s -> csi grpc call -> democratic-csi running privileged on node

In the first case, the nodejs code in this project would need code to directly invoke windows' cli tools etc.

In the second case, the nodejs code in this project would need to invoke the csi-proxy grpc calls which in turn would invoke the windows sys calls etc.

I'm inclined to go the former route as that appears to be the 'blessed' way (and we inherit all the mount/format/etc logic from the proxy). However, at the moment it appears the csi-proxy only supports smb and doesn't do iscsi :(

**Comment by batiati** [comment]
>Well, if I'm understanding things correctly (and your proposed solution) the named pipe won't help because the node container (on windows) cannot run in privileged mode. 

Yeah, you're right

Windows does not have "privileged mode", in fact, neither "isolated mode" too ... windows containers can do a lot of privileged things (some times, the isolation seems to be more a kind of limitation than desigin choice!)

So, my first guess was running the iSCSI powershell's cmdlet from the container, but it worked only for commands that does not depend on MSiSCSI service, as this service can't run inside a container. Even if it could run, I think that we would have some throuble mounting the disk on a global mount-point, there is not like `SmbGlobalMapping` for iSCSI.

(...)

Then, I tried to perform a remote session from-container-to-host, like that:

```
$hostName = "some-address"
$user = "some-credential"
$password = "some-password"
winrm set winrm/config/client "@{TrustedHosts=""$hostName""}" 

[securestring]$secStringPassword = ConvertTo-SecureString $password -AsPlainText -Force
[pscredential]$credential = New-Object System.Management.Automation.PSCredential ($user, $secStringPassword)

$session = New-PSSession -ComputerName $hostName -Credential $credential
Invoke-Command -Session $session -ScriptBlock { Get-IscsiTarget } 
Remove-PSSession -Session $session
```

It worked, but it has **HUGE** implications and security problems ... Really it's not a good design that is worthy to put effort on it. 

(...)

>I'm inclined to go the first route as that appears to be the 'blessed' way (and we inherit all the mount/format/etc logic from the proxy). However, at the moment it appears the csi-proxy only supports smb and doesn't do iscsi :(

I took a quick look at CSIProxy code, it's a considerable task to define an implement all required iSCSI interfaces, but it's possible to do...

I think that there is no need to implement native bindings to `iscsidsc.dll`. Most of CSIProxy calls are implemented through **powershell invoke**, iSCSI could be done in the same way. Complex scenarios like multi-path can be avoided, just as FlexVolume does.

![image](https://user-images.githubusercontent.com/5728002/92251452-43a22080-eea3-11ea-8f7e-7a8e70c51207.png)

(...)

### Making it all simpler

I have a workarround for my specific use-case: Is it possible to generate deterministic names for pvc? Something like [freenas-provisioner ](https://github.com/nmaupu/freenas-provisioner/blob/master/deploy/class.yaml) have:

 ```
  # if enabled created datasets will adhere to reliable pattern
  # if datasetNamespaces == true dataset pattern is: <datasetParentName>/<namespace>/<PVC Name>
  # if datasetNamespaces == false dataset pattern is: <datasetParentName>/<namespace>-<PVC Name>
  # if disabled, datasets will be created with a name pvc-<uid> (the name of the provisioned PV)
  # default: true
  #datasetDeterministicNames:
```
 
So, I can use the legacy FlexVolume iSCSI to connect without having to discover the IQN based on the pvc name.


Best Regards!

**Comment by batiati** [comment]
For closure's sake:

### iSCSI

- There is no support for iSCSI on [Windows CSI-Proxy](https://github.com/kubernetes-csi/csi-proxy)

- The workarround for deterministic IQN names worked well, it's possible to provision a volume using democratic-csi and bind it to the pod using FlexVolume

Be aware of using **StatefullSet** when binding iSCSI disks on Windows; Using Deployment causes Windows to crash when recreating/scaling the pods ... 

FlexVolume couldn't mount volumes created from snapshots ... I did not investigate it yet.

Example of iSCSI binding
```
      volumes:
      - flexVolume:
          driver: microsoft.com/iscsi.cmd
          fsType: NTFS
          options:
            authType: NONE
            chapAuthDiscovery: "false"
            chapAuthSession: "false"
            iqn: iqn.2005-10.org.freenas.ctl:default-poc01
            lun: "0"
            targetPortal: 192.168.1.1:3260
          secretRef: {}
        name: data
```

### SMB

- Its possible to use the FlexVolume in the same way for SMB, using deterministics share names.

There is some issue regarding SMB guest access that I couldn't figure out, even using username and password to connect to the share. So, the following commands are required to run on all nodes to enable SMB guest access:

```
Set-ItemProperty HKLM:\SYSTEM\CurrentControlSet\Services\LanmanWorkstation\Parameters AllowInsecureGuestAuth -Value 1 
restart-service LanmanWorkstation -Force 
```

Snapshots and cloned volumes works great

There is an example of SMB binding, be aware of creating the required [secret](https://github.com/microsoft/K8s-Storage-Plugins/blob/master/flexvolume/windows/sample_yamls/smb-secret.yaml) with username and password.

```
      volumes:
      - flexVolume:
          driver: "microsoft.com/smb.cmd"
          options:
            source: \\192.168.1.1\default-smb01
          secretRef:
            name: smb-secret
        name: data
```

That's it,
Thanks a lot!
